### PR TITLE
docs: Document TPM-backed rootfs encryption

### PR DESCRIPTION
**Pull request by pothos**
A long-requested feature is disk encryption. In the next Alpha we have everything in place for TPM-backed disk encryption with systemd-cryptenroll and Clevis, and with a network-backed disk secret store with Tang.
Document the limitations, implications and the Ignition configs.

Preview URL suffix: `/docs/latest/setup/security/luks/`

### CODE REVIEW

**Comment by pothos** [comment]
I've added a helper service for stronger PCR binding when updates are disabled

**Comment by pothos** [comment]
> I've added a helper service for stronger PCR binding when updates are disabled

Also added a more advanced setup that can do rebinding on updates.
While testing I noticed that the current use of GRUB in Flatcar doesn't measure anything in PCR 8, we should fix that to make stronger binding possible.

**Comment by pothos** [comment]
> While testing I noticed that the current use of GRUB in Flatcar doesn't measure anything in PCR 8, we should fix that to make stronger binding possible.

What I found was `insmod tpm` and `tpm_record_pcrs 0-9` but it sometimes doesn't directly work and I didn't find any docs on that…

File: content/docs/latest/setup/security/luks.md [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+```yaml
+variant: flatcar
+version: 1.0.0
+storage:
+  files:
+    - path: /oem/grub.cfg
+      append:
+        - inline: |
+            insmod tpm
+            tpm_record_pcrs 8-9
```
Comment (pothos) [inline content/docs/latest/setup/security/luks.md:113]: I think `tpm_record_pcrs` is not needed, but not really sure yet.

Comment (jepio) [reply content/docs/latest/setup/security/luks.md:113]: i would remove this and make sure grub does the measuring by default. we don't want everyone to run a different configuration that we risk breaking

**Comment by pothos** [comment]
Without PCR binding the updates don't need to be disabled. The last two examples use PCR binding and there we have one disabling updates and while the other has a helper service to rebind. The rebinding does not handle all possible cases and we could warn about that if people, for example, edit `grub.cfg` and then reboot without running `/oem/bin/oem-postinst`. As recommendation we could say that one should copy the recovery out or maybe better set up a PIN for serial console unlocking.
I think we can get this documented with warnings because even with signed policies it's still possible to lock oneself out (e.g., firmware update) and a backup plan is needed.

File: content/docs/latest/setup/security/luks.md (hunk 1 above)
Comment (pothos) [reply content/docs/latest/setup/security/luks.md:113]: I prepared a change here: https://github.com/flatcar/scripts/pull/1861

**Comment by pothos** [comment]
Do you think it makes sense to test the PCR binding in mantle? When we improve things in Flatcar to have less changes in the boot code path it wouldn't break the helper services but only make the additional reboots unnecessary.

**Comment by jepio** [comment]
yes, we definitely need PCR binding tests in mantle too

File: content/docs/latest/setup/security/luks.md [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        [Install]
+        WantedBy=multi-user.target
+```
+
+Since we normally want updates you have to prevent locking yourself out by removing the PCR binding before updating and add it back after two reboots (a second reboot, after one minute of uptime in the updated system, is needed to create a stable GRUB code path where no GPT flags change). The following example shows how to automate this with a helper that runs systemd-cryptenroll as post-update hook.
+
+Example for TPM2-backed rootfs encryption with systemd-cryptenroll and stronger PCR binding (requires UEFI), with a added unbinding while the update reboot is pending (Butane YAML):
+
+```yaml
+variant: flatcar
```
Comment (ader1990) [inline content/docs/latest/setup/security/luks.md:168]: tested also on Hyper-V with TPM2 functionality, works as expected.

**Review by ader1990 — DISMISSED** [review]

**Comment by pothos** [comment]
By removing the PCR 5 binding the additional update reboot can be removed and I think this has no security implications.

**Comment by pothos** [comment]
> yes, we definitely need PCR binding tests in mantle too

Tests added in https://github.com/flatcar/mantle/pull/521

I also think we could try to unify the PCR values for the first boot and the other boots by not using a kernel parameter for first-boot or not but measure the effective Ignition config on every boot.

**Comment by pothos** [comment]
> [](https://github.com/flatcar/mantle/pull/521)
> 
> I also think we could try to unify the PCR values for the first boot and the other boots by not using a kernel parameter for first-boot or not but measure the effective Ignition config on every boot.

That won't happen for this release, I'll merge it as is now and hope we can get rid of the first reboot soon.

**Pull request merged**